### ISSUE TITLE
how do i use cWRP with Instantsearch widget?

### ISSUE THREAD

**Issue by harshmaur** [algolia/instantsearch.js#1618]
I have a page which contains links that open the same page again but with a different query, since the component remains the same, react will not unmount and mount back again.

I want the parameters to change when the same page is opened up with a different query.

Basically I want to trigger change of state programmatically in my cWRP.

I am using react-instantsearch beta 18

**Comment by vvo** [comment]
Hi @harshmaur, what is "cWRP"?

You should be able to inject state on your `<InstantSearch/>` instance and update the query accordingly, see https://community.algolia.com/instantsearch.js/react/guides/advanced-topics.html#url-routing

**Comment by harshmaur** [comment]
Sorry, its `componentWillReceiveProps`. 

I have a case like this. 

https://cl.ly/0m2y1r091N3i

I have a page which has instantsearch widget loaded in. Below I have similar products, which open up the same page only. So ultimately react does not unmount the component and nothing is changed. For that I use `cWRP` to check for change of any route and change the state.

**Comment by vvo** [comment]
@harshmaur Do you need to wrap this product page as a whole in an `<InstantSearch>`? I would say only the relevant product part needs to be udpated.

As for the page not being updated I fail at seeing how this is a react-instantsearch issue? I would first try to do it without using `<InstantSearch/>` and bind similar product to static data and see if you are able to update the page. This look more of a generic react routing issue rather than a react-instantsearch issue?

**Comment by harshmaur** [comment]
My whole page contains two instantsearch instances. Everything else other than Similar products are in one instance and similar is in another. My first instance basically looks at the url and fetches the route to put some query in filters. On clicking similar product, my route changes, but Instantsearch would not know about it as it sees there is literally no change. Its not a react-instantsearch problem but a react problem. For that react provides cWRP which says if there is a change in the prop, do something that you want to do and update other components. 

So basically I want to know how can I update instantsearch instance to take the new route and render accordingly.

**Comment by mthuret** [comment]
@harshmaur can you provide a jsbin/small repo with a simplified example showing us the issue ? 

In order to have the InstantSearch instance updated you need to pass a new `state` props. I assume that because the route is changing, then the `componentWillReceiveProps` of the component to render will be triggered. Then you need to parse the url to reconstruct the state, and use that state to update the InstantSearch instance.

**Comment by harshmaur** [comment]
I am not doing any url syncs. I am normally getting the slugs from `this.props.params.slug` to find slug and use this in instantsearch like this in my `render()`

```jsx
<InstantSearch appId={ALGOLIA_APPID}
                apiKey={ALGOLIA_APIKEY}
                indexName={LOWESTINDEX}
                searchParameters={{
                    distinct: false,
                    filters: `slug: ${slug} AND instock: 1`
                }}>
```

When this slug changes, I need to get the new slug. 

```jsx
componentWillReceiveProps(nextProps) {
    if(this.props.params.slug !== nextProps.params.slug){
      
   // Change the state here.... and have Instantsearch take the new slug. 
    }
  }
```

I am not sure if I can provide a working jsbin example or not. If the above description is not clear, I can try to do it.

**Comment by mthuret** [comment]
Behavior: if the `searchParameters` props is updated, nothing will happen because `searchParameters` are not updated internally. 

There's two things here: 

* we never update the `searchParameters` property once `InstantSearch` is mounted. It means that a user will never be able to change them once `InstantSearch` is mounted. 
* when `InstantSearch` is uncontrolled (meaning that the state is not handle externally), we do not perform a new search. So even if we were updating the `searchParameters`, nothing would happen.

**Solutions to achieve this use-case**: add a VirtualWidget with the `slug` param as the defaultRefinement.  

@vvo, @bobylito: wdyt about this? If we don't allow the searchParameters to be updated, maybe we should be clear about that in our documentation. Also, VirtualWidgets can be used in many different ways, I guess we can improve the guide about them :)

**Comment by vvo** [comment]
> Solutions to achieve this use-case: add a VirtualWidget with the slug param as the defaultRefinement.

This solution looks like a workaround (good one). But still in react I would expect that updating any prop on InstantSearch would actually do what I want: update the corresponding linked behaviors. What do you think?

Ultimately we might just have a state with:

```
searchParameters: {
  page: 0,
  filters: ...
  distinct: ...
},
widgets: {
  menu: {
    brands: 'Sony'
  }
}
```

And users will be able to update any searchParameters (sorry this is confusing, I am not really talking about SearchParameters from the helper here, just all API parameters that are not handled by widgets.

**Comment by mthuret** [comment]
Well I don't disagree, I think you expect `searchParameters` to update. Also, it's the case with all the other props (algoliaClient, apiKey...). 

One thing we need to check is: what are the side effects if we perform a search from `componentWillReceiveProps` for an uncontrolled `InstantSearch`.

**Comment by vvo** [comment]
> One thing we need to check is: what are the side effects if we perform a search from componentWillReceiveProps for an uncontrolled InstantSearch.

I have no idea for now what is the effect
